apps/omg_status/lib/omg_status/monitor/memory_monitor.ex

```diff
@@ -0,0 +1,167 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.Status.Monitor.MemoryMonitor do
+  @moduledoc """
+  Monitors and raises the :system_memory_too_high alarm when the system memory reaches
+  the specified threshold.
+
+  Intentionally raising a different alarm name from :memsup (which uses :system_memory_high_watermark)
+  so there is no ambiguity to which module is responsible for which alarm.
+
+  See http://erlang.org/pipermail/erlang-questions/2006-September/023144.html
+  """
+  use GenServer
+  require Logger
+
+  @type t :: %__MODULE__{
+          alarm_module: module(),
+          memsup_module: module(),
+          interval_ms: pos_integer(),
+          threshold: float(),
+          raised: boolean(),
+          tref: reference() | nil
+        }
+
+  defstruct alarm_module: nil,
+            memsup_module: nil,
+            interval_ms: nil,
+            threshold: 1.0,
+            raised: false,
+            tref: nil
+
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  # monitor init
+  def init([_ | _] = opts) do
+    _ = Logger.info("Starting #{inspect(__MODULE__)}.")
+    install_alarm_handler()
+
+    alarm_module = Keyword.fetch!(opts, :alarm_module)
+    memsup_module = Keyword.fetch!(opts, :memsup_module)
+    interval_ms = Keyword.fetch!(opts, :interval_ms)
+    threshold = Keyword.fetch!(opts, :threshold)
+
+    state = %__MODULE__{
+      alarm_module: alarm_module,
+      memsup_module: memsup_module,
+      interval_ms: interval_ms,
+      threshold: threshold
+    }
+
+    {:ok, state, {:continue, :first_check}}
+  end
+
+  # gen_event init
+  def init(_args) do
+    {:ok, %{}}
+  end
+
+  # We want the first check immediately upon start, but we cannot do it while the monitor
+  # is not fully initialized, so we need to trigger it in a :continue instruction.
+  def handle_continue(:first_check, state) do
+    _ = send(self(), :check)
+    {:noreply, state}
+  end
+
+  def handle_info(:check, state) do
+    exceed_threshold? = system_memory_exceed_threshold?(state.memsup_module, state.threshold)
+    _ = raise_clear(state.alarm_module, state.raised, exceed_threshold?)
+
+    {:ok, tref} = :timer.send_after(state.interval_ms, :check)
+    {:noreply, %{state | tref: tref}}
+  end
+
+  def handle_cast(:set_alarm, state) do
+    {:noreply, %{state | raised: true}}
+  end
+
+  def handle_cast(:clear_alarm, state) do
+    {:noreply, %{state | raised: false}}
+  end
+
+  #
+  # gen_event handlers
+  #
+  def handle_call(_request, state), do: {:ok, :ok, state}
```

Comment:
Why not `:reply` instead of `:ok`?

Reply:
This is a :gen_event callback so the pattern is different from the GenServer 's `handle_call/3` callback. See https://erlang.org/doc/man/gen_event.html#Module:handle_call-2